mono/mini/aot-runtime.c

```diff
@@ -1947,10 +1947,14 @@ init_amodule_got (MonoAotModule *amodule)
 	int i, npatches;
 
 	/* These can't be initialized in load_aot_module () */
-	if (amodule->shared_got [0] || amodule->got_initializing)
+	if (amodule->got_initialized)
 		return;
 
-	amodule->got_initializing = TRUE;
+	amodule_lock (amodule);
+	if (amodule->got_initialized) {
```

Comment:
This is a low level lock, it cannot be used here, use the loader lock instead.

Reply:
Also, since this function is idempotent, it might be possible to implement it using memory barriers etc.

Reply:
@vargaz Very good points. Went for a single barrier so the change is as less intrusive as possible.